keps/sig-scheduling/3370-native-podgroup-api/README.md

```diff
@@ -0,0 +1,1096 @@
+
+	// Running is the number of pods targeted by this subset
+	// with a Running phase.
+	Running int32
+}
+```
+
+ After the pod group is scheduled for the first time, the condition `PodGroupScheduled` is
+ added by the `pod-group-controller` in kube-controller-manager. Then `pod-group-controller`
+ will reconcile the status of pod group. The workload operator can watch the pod group status,
+ for example, when a pod is deleted and the total running count is less than `MinMember`. The
+ workload operator can determine if it needs to evict the whole group or just create a new pod.
+
+
+### User workflow
+1. The user creates a PodGroup object.
+2. The user creates a workload (Deployment, Job, etc.) and associates .spec.podGroup to the previously created PodGroup name.
+3. The user watched the status of PodGroup. Pods associated with this PodGroup are expected to be co-scheduled if resources are adequate.
+4. The user waits for the PodGroup to be complete and then deletes it.
+
+### Scheduler
+#### QueueSort
+In order to maximize the chance that the pods which belong to the same `PodGroup` to be scheduled consecutively, we need to implement a customized `QueueSort` plugin to sort the Pods properly.
+
+Firstly, we will inherit the default in-tree PrioritySort plugin so as to honor .spec.priority to ensure high-priority Pods are always sorted ahead of low-priority ones.
+
+Secondly, if two Pods hold the same priority, the sorting precedence is described as below:
+
+- If they are both regularPods (without pod.spec.podGroup), compare their 
+`InitialAttemptTimestamp` field: the Pod with earlier `InitialAttemptTimestamp` is positioned 
+ahead of the other.
+  
+- If one is regularPod and the other is pgPod (with pod.spec.podGroup), compare regularPod's
+`InitialAttemptTimestamp` with the `creationtime` of pgPod's PodGroup: the Pod with earlier 
+timestamp is positioned ahead of the other.
+  
+- If they are both pgPods:
+  - Compare their pod group's `CreationTimestamp`: the Pod in pod group which has earlier timestamp is positioned ahead of the other.
+  - If their `CreationTimestamp` is identical, order by their UID of PodGroup: a Pod with lexicographically greater UID is scheduled ahead of the other Pod. (The purpose is to tease different PodGroups with the same `CreationTimestamp` apart, while also keeping Pods belonging to the same PodGroup back-to-back)
+
+#### PreFilter
+In PreFilter phase, we need to check if pod.spec.podGroup is nil firstly. If so, return 
+success. If not, check if the pod group exists in the scheduler's internal cache and pod.spec.
+podGroup.subset is matched in pod group, if not, reject the pod in PreFilter.
+
+
+#### Permit
+In Permit phase, It is divided into two parts.
+
+- The first part is that we put the pods that doesn't meet MinMember into the WaitingMap and reserve resources until MinMember are met or timeout is triggered.
+
+   - Get the number of Running pods that belong to the same PodGroup
```

Comment:
Get by labelSelector? And where's the labelSelector located, in PodGroup level or SubSet level?

Reply:
I ask this question for I find in `SubsetStatus`, we will get the pods by label selector too.

Reply:
Not use label selector, we add a filed called `podgroup` in pod.spec